File: objc/extension/SFDragSprite.m
//
//  SFDragSprite.m
//  SpriteForest2
//
//  Created by Moky on 12-10-3.
//  Copyright 2012 Slanissue.com. All rights reserved.
//

#import "SFDragSprite.h"


@implementation SFDragSprite

@synthesize dragRange = dragRange_;
@synthesize dragTied = dragTied_;

- (id) init
{
	if ((self = [super init]))
	{
		self.dragRange = CGRectZero;
		self.dragTied = NO;
	}
	return self;
}

- (void) onEnter
{
	controller_.dragEnabled = YES;
	[super onEnter];
}

- (void) setAttributes:(NSDictionary *)dict
{
	[super setAttributes:dict];
	
	// whether limit the draging in a range
	NSString * range = [dict objectForKey:@"DragRange"];
	if ([range isKindOfClass:[NSString class]])
	{
		CNLog(@"drag range = %@", range);
		self.dragRange = CGRectFromString(range);
	}
	
	// whether fly back to began point after drag end
	self.dragTied = [[dict objectForKey:@"DragTied"] boolValue];
}

@end


@implementation SFTouchHandler (drag)

- (BOOL) onDragStart:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onDragStart" node:node];
	[pool release];
	return YES;
}

- (BOOL) onDrag:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onDrag" node:node];
	[pool release];
	
	// check range
	if ([node isKindOfClass:[SFDragSprite class]])
	{
		CGRect range = ((SFDragSprite *)node).dragRange;
		if (!CGRectEqualToRect(range, CGRectZero))
		{
			CGSize size = node.contentSize;
			CGPoint anchor = ccpCompMult(ccpFromSize(size), node.anchorPoint);
			CGPoint point = ccpSub(node.position, anchor);
			// x
			if (point.x < range.origin.x)
			{
				point.x = range.origin.x;
			}
			else if (point.x + size.width > range.origin.x + range.size.width)
			{
				point.x = range.origin.x + range.size.width - size.width;
			}
			// y
			if (point.y < range.origin.y)
			{
				point.y = range.origin.y;
			}
			else if (point.y + size.height > range.origin.y + range.size.height)
			{
				point.y = range.origin.y + range.size.height - size.height;
			}
			// limit position
			point = ccpAdd(point, anchor);
			if (!CGPointEqualToPoint(point, node.position))
			{
				node.position = point;
			}
		}
	}
	
	return YES;
}

- (BOOL) onDragEnd:(CCNode *)node
{
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
	[self doEvent:@"onDragEnd" node:node];
	[pool release];
	
	if ([node isKindOfClass:[SFDragSprite class]])
	{
		return !((SFDragSprite *)node).dragTied;
	}
	
	return YES;
}

@end

File: spriteforest-x/extension/SFDragSprite.cpp
//
//  SFDragSprite.cpp
//  SpriteForest-X
//
//  Created by Moky on 12-12-8.
//  Copyright 2012 Slanissue Inc. All rights reserved.
//

#include "SFDragSprite.h"

NS_SF_BEGIN

USING_NS_CC;
USING_NS_CN;

SFDragSprite::SFDragSprite(void)
: SFTouchSprite()
, m_tDragRange(CCRectZero)
, m_bDragTied(false)
{
	
}

SFDragSprite::~SFDragSprite(void)
{
	
}

bool SFDragSprite::init(void)
{
	if (!SFTouchSprite::init())
	{
		CNLog("error");
		return false;
	}
	
	CNTouchController * pController = getController();
	if (pController)
	{
		pController->setDragEnabled(true);
	}
	
	m_tDragRange = CCRectZero;
	m_bDragTied = false;
	
	return true;
}

bool SFDragSprite::setAttributes(CCDictionary * pDict)
{
	if (!SFTouchSprite::setAttributes(pDict))
	{
		CNLog("error");
		return false;
	}
	
	// DragRange
	CCString * pDragRange = (CCString *)pDict->objectForKey("DragRange");
	m_tDragRange = pDragRange ? CCRectFromString(pDragRange->getCString()) : CCRectZero;
	
	// DragTied
	CCString * pDragTied = (CCString *)pDict->objectForKey("DragTied");
	m_bDragTied = pDragTied ? pDragTied->boolValue() : false;
	
	return true;
}

void SFDragSprite::appointHandler(CCDictionary * pEventTree, CCDictionary * pActionTree)
{
	SFDragHandler * pHandler = SFDragHandler::handler(pEventTree, pActionTree);
	setEventDelegate(pHandler);
}

#pragma mark - SFDragHandler

SFDragHandler * SFDragHandler::handler(CCDictionary * pEventTree, CCDictionary * pActionTree)
{
	SFDragHandler * pHandler = new SFDragHandler();
	if (pHandler && pHandler->init(pEventTree, pActionTree))
	{
		pHandler->autorelease();
		return pHandler;
	}
	else
	{
		CC_SAFE_RELEASE(pHandler);
		return NULL;
	}
}

bool SFDragHandler::onDragStart(CCNode * pNode)
{
	CCPoolManager * pm = CCPoolManager::sharedPoolManager();
	pm->push();
	doEvent("onDragStart", pNode);
	pm->pop();
	return true;
}

bool SFDragHandler::onDrag(CCNode * pNode)
{
	CCPoolManager * pm = CCPoolManager::sharedPoolManager();
	pm->push();
	doEvent("onDrag", pNode);
	pm->pop();
	
	// check range
	SFDragSpriteProtocol * pDragSprite = dynamic_cast<SFDragSpriteProtocol *>(pNode);
	if (pDragSprite)
	{
		CCRect range = pDragSprite->getDragRange();
		if (!CCRect::CCRectEqualToRect(range, CCRectZero))
		{
			CCSize size = pNode->getContentSize();
			CCPoint anchor = pNode->getAnchorPointInPoints();
			CCPoint position = pNode->getPosition();
			CCPoint point = ccpSub(position, anchor);
			// x
			if (point.x < range.origin.x)
			{
				point.x = range.origin.x;
			}
			else if (point.x + size.width > range.origin.x + range.size.width)
			{
				point.x = range.origin.x + range.size.width - size.width;
			}
			// y
			if (point.y < range.origin.y)
			{
				point.y = range.origin.y;
			}
			else if (point.y + size.height > range.origin.y + range.size.height)
			{
				point.y = range.origin.y + range.size.height - size.height;
			}
			// limit position
			point = ccpAdd(point, anchor);
			if (!CCPoint::CCPointEqualToPoint(point, position))
			{
				pNode->setPosition(point);
			}
		}
	}
	
	return true;
}

bool SFDragHandler::onDragEnd(CCNode * pNode)
{
	CCPoolManager * pm = CCPoolManager::sharedPoolManager();
	pm->push();
	doEvent("onDragEnd", pNode);
	pm->pop();
	
	SFDragSpriteProtocol * pDragSprite = dynamic_cast<SFDragSpriteProtocol *>(pNode);
	if (pDragSprite)
	{
		return !pDragSprite->getDragTied();
	}
	
	return true;
}

NS_SF_END
